# Random Forest 
### (untuk masuk ke bagian ini, lebih baik pelajari decision tree)

##### Random forest adalah algoritma machine learning. Algoritma ini hampir sama dengan decision tree. Perbedaannya pada decision tree, pohon yang dibangun hanya satu, sedangkan pada random forest diciptakan beberapa pohon. Setiap pohon saling mengeluarkan hasilnya sendiri. Hasil yang diambil adalah hasil yang paling banyak muncul.

Bayangkan anda berada di sebuah hutan. Hutan ini adalah hutan ajaib, masing-masing pohon memiliki kemampuan spiritual sendiri seperti membaca suhu dan membaca cuaca.

- Ketika anda ingin keluar anda bertanya kepada hutan ini "bagaimana cuaca hari ini?".
- Pohon dengan kemampuan membaca cuaca akan membaca cuaca, kemudian masing-masing pohon mengeluarkan hasil, tetapi karena hasil yang berbeda maka pohon ini berdiskusi terlebih dahulu. Pohon A mengatakan hari ini hujan, Pohon B mengatakan hari ini cerah, Pohon C mengatakan hari ini hujan, Pohon D mengatakan hari ini mendung.
- Karena banyak pohon yang meramal bahwa hari ini akan hujan, maka keputusan akhir adalah hari ini akan hujan.
- Ketua pohon akan memberi tahu anda bahwa "hari ini hujan".

begitupun ketika anda bertanya "bagaimana suhu hari ini?", pohon dengan kemampuan membaca suhu akan membaca dan berdiskusi kemudian menyampaikan kepada anda suhu hari ini.

# 1. Penggunaan file

In [1]:
import pandas as pd
import numpy as np
import sklearn.tree as tree
import sklearn.model_selection as ms
import sklearn.ensemble as ens
import sklearn.metrics as met
rf = ens.RandomForestClassifier(n_estimators=100)

try:
    df1 = pd.read_csv("adult_data.csv")
    print("File berhasil dibaca")
except fileNotFoundError:
    print("File tidak berhasil dibaca")
    
df1.head(10)

File berhasil dibaca


age          workclass  fnlwft   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital_status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3      Married-civ-spouse   Handlers-cleaners         Husband   Black   
4      Married-civ-spouse      Prof-specialty            Wife   Black   
5      Married-civ-spouse     Exec-managerial            Wife   White   
6   Married-spouse-absent       Other-service   Not-in-family   Black   
7      Married-civ-spouse     Exec-managerial         Husband   White   
8           Never-married      Prof-specialty   Not-in-family   White   
9      Married-civ-spouse     Exec-managerial         Husband   White   

       sex  capital_gain  capital_loss  hours-per-week  native_country  income  
0     Male          2174             0              40   United-States   <=50K  
1     Male             0             0              13   United-States   <=50K  
2     Male             0             0              40   United-States   <=50K  
3     Male             0             0              40   United-States   <=50K  
4   Female             0             0              40            Cuba   <=50K  
5   Female             0             0              40   United-States   <=50K  
6   Female             0             0              16         Jamaica   <=50K  
7     Male             0             0              45   United-States    >50K  
8   Female         14084             0              50   United-States    >50K  
9     Male          5178             0              40   United-States    >50K

In [12]:
# proses encoding dengan sklearn.preprocessing

import sklearn.preprocessing as prep
gender = prep.LabelEncoder()
gender.fit(df1["sex"])
df1["sex_code"] = gender.transform(df1["sex"])
df1[["sex", "sex_code"]].head(10)

education = prep.LabelEncoder()
education.fit(df1["education"])
df1["education_code"] = education.transform(df1["education"])
df1[["education", "education_code"]].head(10)

race = prep.LabelEncoder()
race.fit(df1["race"])
df1["race_code"] = race.transform(df1["race"])
df1[["race", "race_code"]].head(10)

workclass = prep.LabelEncoder()
workclass.fit(df1["workclass"])
df1["workclass_code"] = workclass.transform(df1["workclass"])
df1[["workclass", "workclass_code"]].head(10)

occupation = prep.LabelEncoder()
occupation.fit(df1["occupation"])
df1["occupation_code"] = occupation.transform(df1["occupation"])
df1[["occupation", "occupation_code"]].head(10)

relationship = prep.LabelEncoder()
relationship.fit(df1["relationship"])
df1["relationship_code"] = relationship.transform(df1["relationship"])
df1[["relationship", "relationship_code"]].head(10)

native_country = prep.LabelEncoder()
native_country.fit(df1["native_country"])
df1["native_country_code"] = native_country.transform(df1["native_country"])
df1[["native_country", "native_country_code"]].head(10)

df1[["sex", "sex_code", "education", "education_code", "race", "race_code", "workclass", "workclass_code", "occupation", "occupation_code", "relationship", "relationship_code", "native_country", "native_country_code"]].head(10)

sex  sex_code   education  education_code    race  race_code  \
0     Male         1   Bachelors               9   White          4   
1     Male         1   Bachelors               9   White          4   
2     Male         1     HS-grad              11   White          4   
3     Male         1        11th               1   Black          2   
4   Female         0   Bachelors               9   Black          2   
5   Female         0     Masters              12   White          4   
6   Female         0         9th               6   Black          2   
7     Male         1     HS-grad              11   White          4   
8   Female         0     Masters              12   White          4   
9     Male         1   Bachelors               9   White          4   

           workclass  workclass_code          occupation  occupation_code  \
0          State-gov               7        Adm-clerical                1   
1   Self-emp-not-inc               6     Exec-managerial                4   
2            Private               4   Handlers-cleaners                6   
3            Private               4   Handlers-cleaners                6   
4            Private               4      Prof-specialty               10   
5            Private               4     Exec-managerial                4   
6            Private               4       Other-service                8   
7   Self-emp-not-inc               6     Exec-managerial                4   
8            Private               4      Prof-specialty               10   
9            Private               4     Exec-managerial                4   

     relationship  relationship_code  native_country  native_country_code  
0   Not-in-family                  1   United-States                   39  
1         Husband                  0   United-States                   39  
2   Not-in-family                  1   United-States                   39  
3         Husband                  0   United-States                   39  
4            Wife                  5            Cuba                    5  
5            Wife                  5   United-States                   39  
6   Not-in-family                  1         Jamaica                   23  
7         Husband                  0   United-States                   39  
8   Not-in-family                  1   United-States                   39  
9         Husband                  0   United-States                   39

In [20]:
df1.head(11)

age          workclass  fnlwft      education  education_num  \
0    39          State-gov   77516      Bachelors             13   
1    50   Self-emp-not-inc   83311      Bachelors             13   
2    38            Private  215646        HS-grad              9   
3    53            Private  234721           11th              7   
4    28            Private  338409      Bachelors             13   
5    37            Private  284582        Masters             14   
6    49            Private  160187            9th              5   
7    52   Self-emp-not-inc  209642        HS-grad              9   
8    31            Private   45781        Masters             14   
9    42            Private  159449      Bachelors             13   
10   37            Private  280464   Some-college             10   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
5       Married-civ-spouse     Exec-managerial            Wife   White   
6    Married-spouse-absent       Other-service   Not-in-family   Black   
7       Married-civ-spouse     Exec-managerial         Husband   White   
8            Never-married      Prof-specialty   Not-in-family   White   
9       Married-civ-spouse     Exec-managerial         Husband   White   
10      Married-civ-spouse     Exec-managerial         Husband   Black   

        sex  ...  hours-per-week  native_country  income sex_code  \
0      Male  ...              40   United-States   <=50K        1   
1      Male  ...              13   United-States   <=50K        1   
2      Male  ...              40   United-States   <=50K        1   
3      Male  ...              40   United-States   <=50K        1   
4    Female  ...              40            Cuba   <=50K        0   
5    Female  ...              40   United-States   <=50K        0   
6    Female  ...              16         Jamaica   <=50K        0   
7      Male  ...              45   United-States    >50K        1   
8    Female  ...              50   United-States    >50K        0   
9      Male  ...              40   United-States    >50K        1   
10     Male  ...              80   United-States    >50K        1   

   education_code  race_code  workclass_code  occupation_code  \
0               9          4               7                1   
1               9          4               6                4   
2              11          4               4                6   
3               1          2               4                6   
4               9          2               4               10   
5              12          4               4                4   
6               6          2               4                8   
7              11          4               6                4   
8              12          4               4               10   
9               9          4               4                4   
10             15          2               4                4   

    relationship_code  native_country_code  
0                   1                   39  
1                   0                   39  
2                   1                   39  
3                   0                   39  
4                   5                    5  
5                   5                   39  
6                   1                   23  
7                   0                   39  
8                   1                   39  
9                   0                   39  
10                  0                   39  

[11 rows x 22 columns]

In [24]:
# Variabel X untuk menampung data yang sudah dibuang data integer (karena kita hanya ingin menggunakand data yang telah dibuat dalam bentuk bilangan)
# Variabel y untuk menampung data tujuan/target. Pada kasus ini target yang akan kita lakukan prediksi adalah "income".
X = df1.drop(["fnlwft", "workclass", "education", "marital_status", "occupation", "relationship", "race", "sex", "native_country", "income"], axis=1)
y = df1["income"]
print(X)

       age  education_num  capital_gain  capital_loss  hours-per-week  \
0       39             13          2174             0              40   
1       50             13             0             0              13   
2       38              9             0             0              40   
3       53              7             0             0              40   
4       28             13             0             0              40   
...    ...            ...           ...           ...             ...   
48837   39             13             0             0              36   
48838   64              9             0             0              40   
48839   38             13             0             0              50   
48840   44             13          5455             0              40   
48841   35             13             0             0              60   

       sex_code  education_code  race_code  workclass_code  occupation_code  \
0             1               9          4  

# 2. Melakukan algoritma random forest

In [4]:
# sisihkan data sebanyak (test_size=0,2) data untuk nantinya sebagai tes apakah prediksi benar atau salah
# setiap 0.1 test_size maka diambil ... data. (dengan rumus (banyak data/10))
# dikasus ini menggunakan 0.2 test_size maka akan diambil ... data. (dengan rumus (banyak data/10))
X_train, X_test, y_train, y_test = ms.train_test_split(X, y, test_size=0.2)

rf.fit(X_train, y_train)

RandomForestClassifier()

In [5]:
# untuk melihat parameter yang terkandung dalam pelatihan model rf.fit(X_train, y_train)
print(rf.get_params())

{'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': 'sqrt', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 100, 'n_jobs': None, 'oob_score': False, 'random_state': None, 'verbose': 0, 'warm_start': False}


In [6]:
y_prediksi = rf.predict(X_test)
print(y_prediksi)

[' <=50K' ' <=50K' ' >50K' ... ' <=50K' ' <=50K' ' <=50K.']


# 3. Memeriksa seberapa akurat Random Forest yang telah dibuat

In [22]:
akurasi = met.accuracy_score(y_test, y_prediksi)
print(akurasi)

0.5072167059064387


In [8]:
print(met.classification_report(y_test, y_prediksi))
print(rf.feature_importances_)

              precision    recall  f1-score   support

       <=50K       0.60      0.74      0.66      4951
      <=50K.       0.31      0.19      0.24      2467
        >50K       0.46      0.47      0.47      1571
       >50K.       0.19      0.12      0.15       780

    accuracy                           0.51      9769
   macro avg       0.39      0.38      0.38      9769
weighted avg       0.47      0.51      0.48      9769

[0.29594666 0.06261884 0.06040222 0.0265953  0.15917754 0.01611303
 0.03914829 0.03342759 0.06623449 0.12165918 0.0823241  0.03635274]


## pada data dapat dilihat adanya:
#### Precision = prediksi positif yang benar dibanding total prediksi positif. Contoh,  prediksi 100 orang pendapatan <=50k tetapi ternyata dari 100 itu hanya 60% yang benar, maka precision 0,60. precision =(jumlah prediksi positif yang benar) / (total prediksi positif).
#### Recall = prediksi positif yang benar dibanding seluruh kelas yang sebenarnya. Contoh, jika sebenarnya ada 150 orang yang pendapatan <=50k dan model hanya bisa mendeteksi 80 orang, maka recall 53,3%.  recall = (jumlah prediksi positif yang benar) / (jumlah keseluruhan anggota kelas yang sebenarnya).
#### F1-score = Rata-rata dari precision dan recall sehingga mencerminkan keseimbangan pada precision dan recall. Semakin tinggi maka semakin baik.
#### Support = Sampel yang masuk dalam kategori yang dipediksi. Dalam hal ini adanya total sampel yaitu 9769.
#### Bagian akhir terdapat data [0,... , .....]. Data ini menunjukkan seberapa pengaruh fitur tersebut terhadap algoritma random forest ini. Semakin besar 0,... maka fitur tersebut semakin penting. Dalam kasus ini fitur-fitur berupa (age, education_num, capital_gain,  capital_loss,  hours-per-week, capitesex_code,	education_code,	race_code,	workclass_code,	occupation_code,	relationship_code,	native_country_code) (12 fitur) 